security/pkg/k8s/controller/workloadsecret.go

```diff
@@ -359,39 +399,54 @@ func (sc *SecretController) scrtUpdated(oldObj, newObj interface{}) {
 	namespace := scrt.GetNamespace()
 	name := scrt.GetName()
 
-	certBytes := scrt.Data[CertChainID]
-	cert, err := util.ParsePemEncodedCertificate(certBytes)
-	if err != nil {
-		log.Warnf("Failed to parse certificates in secret %s/%s (error: %v), refreshing the secret.",
-			namespace, name, err)
-		if err = sc.refreshSecret(scrt); err != nil {
-			log.Errora(err)
+	_, waitErr := sc.certUtil.GetWaitTime(scrt.Data[CertChainID], time.Now(), sc.minGracePeriod)
+
+	caCert, _, _, rootCertificate := sc.ca.GetCAKeyCertBundle().GetAllPem()
+
+	// Check if root certificate in key cert bundle is not up-to-date. With mutiple
+	// Citadel deployed in Istio, and Citadels are in self signed mode, the root
+	// certificate in istio-ca-secret could be rotated by any Citadel and become newer
+	// than the one in local key cert bundle.
+	// When Citadel is in plugged cert mode, the root cert in keycertbundle and root
+	// cert in istio-ca-secret are always identical.
+	if !bytes.Equal(rootCertificate, scrt.Data[RootCertID]) {
+		caSecret, scrtErr := sc.caSecretController.LoadCASecretWithRetry(CASecret,
+			sc.istioCaStorageNamespace, 100*time.Millisecond, 5*time.Second)
```

Comment:
My concern is that this call is triggered for every Istio workload secret when the root cert is rotated and the workload secrets' root certs need to be updated. This is a burst of I/O. This even happens when there's a single Citadel.
We have two solutions:
1. We could maintain a variable "lastKCBSyncTime" to record the last timestamp the KCB synced with the CA secret. And we consider within 30 sec or so, the last-fetched KCB is the same with the CA root secret. We assume the root cert won't be changed twice within 30 sec, which I think is reasonable.
2. The alternative is have a listener on the CA root cert secret, and modify the local KCB once the secret is changed. But if the apiserver is not available during the CA root cert update, the KCB will be temporarily out-of-sync with the root secret. This only happens when you have multiple Citadels, though.
Either way should be OK. If the corner case happens, the workloads will keep being hot-restarted for some time.
 It seems (1) is easier in terms of implementation. We just need to introduce a variable.
WDYT?

Reply:
Actually I prefer (1). (2) seems pretty bad if the corner case happens.

Reply:
This makes sense to me. Thanks. I have implemented option (1).